src/design_docs/async_stack_traces.md

```diff
@@ -0,0 +1,705 @@
+  58:     0x7ff6d4165f16 - std::rt::lang_start_internal::closure$2
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\rt.rs:45
+  59:     0x7ff6d4165f16 - std::panicking::try::do_call
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:403
+  60:     0x7ff6d4165f16 - std::panicking::try
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:367
+  61:     0x7ff6d4165f16 - std::panic::catch_unwind
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panic.rs:129
+  62:     0x7ff6d4165f16 - std::rt::lang_start_internal
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\rt.rs:45
+  63:     0x7ff6d40c45bf - std::rt::lang_start<tuple$<> >
+                               at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\std\src\rt.rs:62
+  64:     0x7ff6d40c12d6 - main
+  65:     0x7ff6d417ad1c - invoke_main
+                               at d:\a01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
+  66:     0x7ff6d417ad1c - __scrt_common_main_seh
+                               at d:\a01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
+  67:     0x7ffbe0a26ab0 - BaseThreadInitThunk
+  68:     0x7ffbe1771dbb - RtlUserThreadStart
+error: process didn't exit successfully: `target\debug\async-std.exe` (exit code: 101)
+```
+
+</details>
+
+## Sync Stack Trace Trimming
+
+Rust supports both a short and full backtraces, controlled by either `RUST_BACKTRACE=1` or `RUST_BACKTRACE=full`. The differents is that short backtraces (`RUST_BACKTRACE=1`) trims away some of the early and late frames.
+
+Below is an example of a short backtrace from a simple program where `main` calls `foo` which calls `bar` which calls `baz` which panics.
+
+```
+thread 'main' panicked at 'explicit panic', src\main.rs:14:5
+stack backtrace:
+   0: std::panicking::begin_panic_handler
+             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:517
+   1: core::panicking::panic_fmt
+             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:101
+   2: core::panicking::panic
+             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:50
+   3: sync::baz
+             at .\src\main.rs:14
+   4: sync::bar
+             at .\src\main.rs:10
+   5: sync::foo
+             at .\src\main.rs:6
+   6: sync::main
+             at .\src\main.rs:2
+   7: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
+             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\ops\function.rs:227
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

Comment:
As you can see regular backtraces are also pretty sub-optimal. IMO, frames 0, 1, 2, and 7 should be elided too. In real backtraces, there are often trampoline functions, etc. which should be elided too.

Reply:
Yeah, I think whatever we do for async stack traces will end up improving sync backtraces quite a bit too.